### ISSUE TITLE
plugin system

### ISSUE THREAD

**Issue by schettino72** [pydoit/doit#32]
Add a plugin system, it should support:

- [x] commands
- [x] reporters
- [x] db-backend
- [ ] runner
- [ ] task configuration?

This plugins could be distributed as separate python packages on PyPI, or just include in the root folder of a project together with `dodo.py`.

**Comment by ankostis** [comment]
I suggest to make use of the [setuptools](https://www.google.it/search?q=setuptools+plugins) plugin mechanism, so that enabling a plugin site-wide is just a matter of: 
`pip install doit_plugin1`.

**Comment by schettino72** [comment]
As I said in the ticket description:

> This plugins could be distributed as separate python packages on PyPI, or just include in the root folder of a project together with dodo.py.

Use setuptools is nice for distributed plugins. But can it work for plugins that are specific for a project and not being distributed? Maybe possible but i didnt see how...

A bigger problem is that not everyone use setuptools!

So I am not saying no to setuptools. It would be nice to support setuptools plugins at some point but it is not my focus right now.

**Comment by Kwpolska** [comment]
Fun fact: you can’t install `doit` without setuptools, and [I am to blame for this](https://github.com/pydoit/doit/commit/de74e34d075823d84db80986238640bc55edf326).

**Comment by schettino72** [comment]
@Kwpolska thats not really true. You can just drop the folder with the source in your python path, or use a distribution package. There are also python packages like conda...  

So while while requiring it on setup.py is ok, requiring it for a plugin system is not. Why you guys didnt use it for Nikola?

**Comment by Kwpolska** [comment]
@ralsina ↑

**Comment by Kwpolska** [comment]
@ralsina 

> So while while requiring setuptools on setup.py is ok, requiring it for a plugin system is not. Why you guys didnt use it for Nikola?

**Comment by ralsina** [comment]
a) I did not know about it

b) I did know about yapsy

**Comment by ankostis** [comment]
Just a note, it wouldn't make sens to give the possibility for plugins to replace "task configuration" unless they can also modify DependencyClass and TaskControl/Dispatcher behavio.

**Comment by ankostis** [comment]
Another item to plugin-ize: *json* Encoders/Decoders

**Comment by schettino72** [comment]
@ankostis Now the plugin use `@` and put the object before the package/module like `MyCmd@mypackage.myplugin`. I think it should be `mypackage.myplugin:MyCmd`...

The problem is that by default `:` is a delimiter for option name and value. In python3 it is easy to configure that [1], not sure how hard it is to do the same in python2. What do you think?

[1] https://docs.python.org/3.4/library/configparser.html#customizing-parser-behaviour

**Comment by saimn** [comment]
There is a package which brings python3's configparser to python2, maybe it can make things easier since python3 version has more features.
https://pypi.python.org/pypi/configparser

**Comment by maxim-s-barabash** [comment]
I do not know how about plug-ins, but add custom commands would not be bad.
As for the reporter, so you can use multiple reporter

dodo.py
    my_cmd1 = MyCmd1(doit.DoitCmdBase): pass
    my_cmd2 = MyCmd2(doit.DoitCmdBase): pass

DOIT_CONFIG = {'default_tasks': ['my_task_1', 'my_task_2'],
 'reporter': ['json', 'my_reporter'],
 'my_cmd': ['my_cmd1', 'my_cmd2'],
}

ankostis, +++

**Comment by ankostis** [comment]
I remember many year ago from my contact with emacs, it has a more "consenting" architecture for extensions, in the sense that the base functionality does not ask pre-specified questions to plugins, _"tell me your new commands"_, _"tell me your own reporter"_, but the plugin decides what to modify.

Of course this means that the base-platform (`doit` then becomes a "platform") has to be designed for extensibility, ie [hooks](http://en.wikipedia.org/wiki/Hooking).  Note that hooks have to be used on equal footing, both from internal code and from plugins.

**Comment by schettino72** [comment]
@ankostis I changed the code to use the backported configparser as suggested by @saimn (thanks).
So please update the plugin to use this style to refer to pluged-in component `mypackage.myplugin:MyCmd`

**Comment by ankostis** [comment]
Done (it was only a matter of documentation).

Now that commands can be plugged-in later, it would be better to make ``doc_description`` a method so as to be possible to override.

**Comment by schettino72** [comment]
@ankostis `Command` has a non-implemented `execute()` method. So it must be subclassed anyway. `doc_description` can be overwritten by a subclass, since it is just some text I see no reason to make it a text.

**Comment by ankostis** [comment]
@schettino72  I'm not sure how to interpret your answer.  Anyway, it is not such a great deal,  I can workaround either by running cmd-stuff at import-time or by eliminating dynamic content from help-strings.  

The problem for plugins is rather more general, in that they do not have the opportunity to "run setup stuff" prior to being invoked for the actual job.

**Comment by schettino72** [comment]
@ankostis Can you give an example of what you are looking for?

**Comment by ankostis** [comment]
Just check [cmd-graphx:line291](https://github.com/pydoit/doit-graphx/blob/master/cmd_graphx.py#L291).  The version-discovery to be imprinted in the help-msg runs at import-time.

**Comment by ankostis** [comment]
The general drawbacks with the current plugin architecture are, in my point of view, the following:


1. We now have 3 sources of setup-data: ``dodo.py``, ``doit.cfg``, ``DOIT_CONFIG``, and particularly having to interact with 2 files, makes things difficult to track, and hence more error prone (i.e. how to configure plugins for IPython's magic-function).

2. Another issue is that plugins don't have the opportunity to "configure themselves", 
that is, they have to do all they need to do in just one step.
 
3. And finally, the way the configuring works is "passive", in the sense that plugins can strictly affect what has been preselected by the platform (doit-base).  
A more liberal approach, with "hooks" (see my above comment), would allow, for instance, to modify existing commands (inherit and replace them).


I'm suggesting the following to address the above:

A. Use *setup_tools* plugin mechanism, which means that plugins are installed site-wide as soon as they are "pip-installed".  

B. Have plugins implement the following "interface":

```
   def plugin_load(conf):
       """
	   Invoked on all plugins, before starting calling their :func:`plugin_install()`.
	   
	   :param dict conf: contains various "hooks" such as commands, dbs, etc
	   """
       pass

   def plugin_install(conf):
       """
	   Invoked after having called :func:`plugin_load()` on all other plugins.
	   
	   :param dict conf: "hooks" as updated from all plugins `plugin_load()`.
	   """
       pass
```
	  
The splitting of initialization in 2-stage avoid inter-plugin races (which loads first).

The contents of the `hooks` dict has to be decided, but a primary candidate is an augmented `DOIT_CONFIG`.

**Comment by schettino72** [comment]
> 1. We now have 3 sources of setup-data: dodo.py, doit.cfg, DOIT_CONFIG

dodo.py and DOIT_CONFIG aren't the same? `dodo.py` is not required by doit! It is used by the default task_loader but event with the default loader not all commands require a `dodo.py`. 

So in the future, yes. DOIT_CONFIG will be redundant to what can be configured in `doit.cfg`. but DOIT_CONFIG wont ever be able to to handle everything `doit.cfg` will do.

> 2. Another issue is that plugins don't have the opportunity to "configure themselves", 

Thats a planned feature. not implemented yet. thats way this issue is still open :)

> 3. And finally, the way the configuring works is "passive".

I considered using hooks, but got the conclusion they are not a good model for doit. Sorry I didnt document my design decisions, I will try to write it down later... 

The problem of modifying existing commands can be easily solved without hooks. For example create new command (you might subclass the original command class) than register it with the same name.

> A. Use setup_tools plugin mechanism, which means that plugins are installed site-wide as soon as they are "pip-installed". 

Second time you ask for setuptools. I already answered that.

> B. Have plugins implement the following "interface":

What problem is this solving?

@ankostis I would prefer we discuss this in the mailing list, writing inline replies on github is very annoying.

**Comment by ankostis** [comment]
(I really like discussion here because of nice formatting)

Moved discussion to forum: (pending approval)

**Comment by schettino72** [comment]
Closing. Plugin system is in place but not all plugin categories implemented, missing categories will come later...
